sematic/ui/packages/common/src/pages/RunDetails/dag/ReactFlowDag.tsx

```diff
@@ -0,0 +1,142 @@
+    const { nodes, onNodesMeasured } = props;
+
+    const scratchPad = useRef<HTMLDivElement>(null);
+
+    useEffect(() => {
+        if (!nodes) {
+            return;
+        }
+
+        const domNodes = scratchPad.current!.querySelectorAll(".node-container");
+
+        // assign initial width to nodes
+        domNodes.forEach((domNode, index) => {
+            const graphNode = nodes[index];
+            graphNode.width = domNode.clientWidth;
+            graphNode.height = 50;
+        });
+
+        onNodesMeasured(nodes);
+    }, [nodes, onNodesMeasured]);
+
+    if (!nodes) {
+        return null;
+    }
+
+    return <div ref={scratchPad} style={{ visibility: "hidden", position: "absolute" }}>
+        {nodes.map((node, index) => {
+            return <NodeContainer className="node-container" key={index}>
+                {node.data.label}
+            </NodeContainer>
+        })}
+    </div>
+}
+
+function ReactFlowDagStaging() {
+    const { graph, isGraphLoading } = useRootRunContext();
+    const { selectedRun } = useRunDetailsSelectionContext();
+
+    const [nodes, setNodes] = useState<Node[] | undefined>(undefined);
+    const [edges, setEdges] = useState<Edge[] | undefined>(undefined);
+    const [version, {inc: incVersion}] = useCounter(0);
+
+    const [layoutDone, setLayoutDone] = useState(false);
+
+    const onNodesMeasured = useCallback((nodes: Node[]) => {
+        layout(nodes, edges!);
+
+        setNodes(nodes);
+        setEdges(edges);
+        setLayoutDone(true);
+        incVersion();
+    }, [setLayoutDone, edges, incVersion]);
+
+    useEffect(() => {
+        if (isGraphLoading) {
+            return;
+        }
+        // isGraphLoading === false
+        setLayoutDone(false);
+        const { nodes, edges } = getReactFlowDag(graph, selectedRun);
+
+        setNodes(nodes);
+        setEdges(edges);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [isGraphLoading]);
+
+    const dagView = useMemo(() => {
+        if (version === 0) {
+            return null;
+        }
+        return <ReactFlowDag nodes={nodes!} edges={edges!} />
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [version]);
+
+    return <>
+        {!layoutDone && <div>
+            <ReactFlowDagScratchPad nodes={nodes} onNodesMeasured={onNodesMeasured} />
+        </div>}
+        {dagView}
+    </>
+}
+
+interface ReactFlowDagProps {
+    nodes: Node[];
+    edges: Edge[];
+}
+
+const nodeTypes = {
+    [NodeTypes.COMPOUND]: CompoundNode,
+    [NodeTypes.LEAF]: LeafNode
+};
+
+function ReactFlowDag(props: ReactFlowDagProps) {
+    const { nodes: _nodes, edges: _edges } = props;
+
+    const [nodes, setNodes, onNodesChange] = useNodesState(_nodes);
+    const [edges, setEdges, onEdgesChange] = useEdgesState(_edges);
+    const reactFlowRef = useRef<ReactFlowInstance>();
+
+    
+    useEffect(() => {
+        setNodes([..._nodes]);
+        setEdges([..._edges]);
```

Comment:
Do you mean to have one dot per each incoming edge? For the run nodes who has a short name, it is a bit hard to arrange many dots on the top border and look nice. 

Currently, the hidden nodes, no matter how many, are removed. So the edges from the hidden nodes and the edges from the source node of the hidden node could be merged as one. So if we do need to support that, it might involve more complicated logic to separate things out. What is the reason that we need to separate them into two dots? I am trying to evaluate the benefits versus the cost. 

| Is it possible to have a tooltip with the argument's name and type on hover over the dot?
Yes. But it might not display well if the (input) artifacts are too many. According to the current Figma design, in this version, we are weakening the presentation of the artifacts on the DAG view. Users should see them from the run detail tabs.

Reply:
I was asking for separate dots so that on mouseover the name and type of that specific argument would be displayed. This would solve the issue when there are many arguments, and would increase readability of the diagram.

Reply:
It is impossible to display many dots for a run with too many arguments. For example, we cannot display 10 dots on the top of the run box for the root pipeline run if the run does have 10 input arguments. There is simply not enough space.

However, if we don't even represent the idea of arguments, like what we are doing now, there will be no such issue to solve.